keps/sig-node/2369-sizing-providers/README.md

```diff
@@ -0,0 +1,363 @@
+	// provider executable as seen by the kubelet. The executable must be in the kubelet's
+	// bin directory (set by the --image-sizing-provider-bin-dir flag).
+	Name string `json:"name"`
+
+	// Required input version of the exec SizingProviderRequest. The returned SizingProviderResponse
+	// MUST use the same encoding version as the input. Supported values are:
+	// - sizingprovider.kubelet.k8s.io/v1alpha1
+	APIVersion string `json:"apiVersion"`
+
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args,omitempty"`
+
+	// Env defines additional environment variables to expose to the process. These
+	// are unioned with the host's environment, as well as variables client-go uses
+	// to pass argument to the plugin.
+	// +optional
+	Env []ExecEnvVar `json:"env,omitempty"`
+}
+
+// ExecEnvVar is used for setting environment variables when executing an exec-based
+// sizing plugin.
+type ExecEnvVar struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+```
+
+### Sizing Provider Request API
+
+If an exec plugin is enabled the kubelet will exec the plugin during the startup, passing the `SizingProviderRequest` API via stdin.
+
+The kubelet will encode the request based on the apiVersion provided in SizingProviderConfig. It wil also exec the plugin based on the `args` and `env` fields in `SizingProviderConfig`.
+
+```go
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SizingProviderRequest will be passed to the plugin via stdin. In general,
+// plugins should prefer responding with the same apiVersion they were sent.
+type SizingProviderRequest struct {
+	metav1.TypeMeta
+}
+```
+
+### Sizing Provider Response API
+
+An exec plugin is expected to return an encoded response of the `SizingProviderResponse` API to the kubelet via stdout. 
+
+```go
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SizingProviderResponse holds sizing values that the kubelet should use for the specified
+// image provided in the original request. Kubelet will read the response from the plugin via stdout.
+type SizingProviderResponse struct {
+	metav1.TypeMeta
+ 
+      // A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G,pid=100) pairs
+	// that describe resources reserved for non-kubernetes components.
+	// Currently only cpu and memory are supported.
+	// See http://kubernetes.io/docs/user-guide/compute-resources for more detail.
+	SystemReserved map[string]string `json:"systemReserve"`
+	// A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G,pid=100) pairs
+	// that describe resources reserved for kubernetes system components.
+	// Currently cpu, memory and local ephemeral storage for root file system are supported.
+	// See http://kubernetes.io/docs/user-guide/compute-resources for more detail.
+	KubeReserved map[string]string `json:"kubeReserve"`
+}
+
+```
+
+### Metrics
+
+N/A
```

Comment:
I think one metric is startup time even though it is not time series, it will be good to know how long kubelet took to startup with a plugin.